### PR TITLE
Rewrite message listing function and headers parsing library

### PR DESCRIPTION
**Pull request by rrao24**
This PR attempts to resolve some issues within message.js and headerParsingLib.js

When querying for a list of messages, a next link is returned regardless of whether or not the following page actually has messages. Messages are also different because they does not use a `page` field; rather, they use `sortKeyLT` to do the pagination managing. The current implementation of headerParsingLib expects there to be a `page` field - so I've rewritten that library to be more generic. To do this, I've replaced the previous header parsing dependency with two core node libraries, url and querystring. Using these libraries allows headerParsingLib.js to remain generic without any 'hacky' coding required.

Some messages tests have been rewritten and the message documentation has been updated as well.

### CODE REVIEW

**Comment by dtolb** [comment]
@rrao24 Can you add a test case covering the edge cases... IE what happens when there is no `next` page. What happens when there are `0` messages. Basically, have the test cases somewhat document the desired behaviour.

**Comment by dtolb** [comment]
This also changes behaviour rolled in by @nguyer ... Can you take a look?

File: test/message-test.js [hunk 1]
```diff
@@ -148,9 +148,10 @@ describe("Message API", function () {
 				.reply(200, messagesList, {
 					"link" : "<https://api.catapult.inetwork.com" +
 						"/v1/users/" + userId + "/messages?fromDateTime=" +
-						fromDateTime + "&" + "toDateTime=" + toDateTime + ">"
```
Comment (nguyer) [inline test/message-test.js:151]: Did we decide not to support querying by date and time at all through the SDK? If we support this API function, maybe we should keep this test, and the sortKey test should be a separate test?

**Comment by nguyer** [comment]
@rrao24 You are correct, the messaging API always returns a `next` link. Perhaps this API should not do that to begin with. If it makes it consistent with the rest of the API/SDK, I'm okay with not returning an empty list at the end. I think that's the way it used to work but that was a while ago I worked on that...

File: test/message-test.js (hunk 1 above)
Comment (rrao24) [inline test/message-test.js:151]: The tests around lines 101 and 112 do test the querying by date and time - these tests were intended to just be pagination tests; I can make that more clear.

**Comment by rrao24** [comment]
@dtolb I've updated the pagination tests as you suggested - there is one which tests whether the pagination mechanism works, and one which tests what would happen on the edge case of there being 0 messages.

@nguyer Making the messages consistent with the rest of the API/SDK is definitely something to consider, but I believe that developers (at least @fuchsnj) were leaning towards making the rest of the SDK/API more like the messages interface. Either way, the code written here should not care about the underlying mechanism for pagination as the `getNextLink` function in `headerParsingLib.js` is written to be generic.

**Comment by dtolb** [comment]
@nguyer @fuchsnj @nbragdon  @a-type 

Looks good to me, one more?

**Comment by a-type** [comment]
Moving from a library to regex makes me uncomfortable, and I don't think I understand why it was done, or what we gain from it here.

**Comment by rrao24** [comment]
@a-type We want the getNextLink function to give us something like `{sortKeyLT : '1'}`, and the parse-link-header library would have given us something like 
`{ next:
   { sortKeyLT : '1',
     rel: 'next',
     url: 'https://api.github.com/user/9287/repos?page=3&per_page=100' }`
  prev:
   { more fields },
  last:
   { more fields} }`.
It is worth noting also that that parse-link-header library does the parsing based off of regex as well. I personally feel that the current implementation is less clunky and fits our use case a bit better, though I could definitely be wrong.

Thank you for your feedback.

**Comment by rrao24** [comment]
I've gone ahead and rewrote the functionality using the parse-link-header library so we can take a look at it and see if we find this to be more reliable.

File: lib/headerParsingLib.js [hunk 2]
```diff
@@ -9,14 +8,11 @@ var querystring = require("querystring");
...
-			if (links[elem].indexOf("rel=\"next\"") !== -1) {
-				var linkRegex = /<(.*)>/;
-				var link = linkRegex.exec(links[elem])[1];
-				var parsedUrl = url.parse(link);
-				return querystring.parse(parsedUrl.query);
-			}
+		var parsedHeader = parse(headers.link);
+		if (parsedHeader.next) {
+			delete parsedHeader.next.rel;
+			delete parsedHeader.next.url;
```
Comment (a-type) [inline lib/headerParsingLib.js]: I'd recommend against ever really using the `delete` keyword (and many popular Javascript style guides agree). If we're really so specific about the returned object that we can't allow these fields to be present, we should be explicit in what we desire by creating a new object, selecting the keys we want.

```
return {
   sortKeyLT: parsedHeader.next.sortKeyLT
};
```

Being explicit protects us against any future changes to the `parse-link-header` library. If the maintainer decides to add a field to the `next` object, we won't have to come back to this code and `delete` it too.

As a rule of thumb in JS, I recommend never mutating objects if it's possible to avoid... really, that can be a good rule for development in general. Or at least, there is a growing trend that is pushing for immutability. It solves a lot of headaches as projects grow in complexity.

**Comment by a-type** [comment]
The reason I prefer using a library (even if the library uses regex internally), is based on two things from my experience:
- Regex in your own code, if not properly encapsulated and thoroughly tested, adds a lot of maintenance cost. Regex always seems like a reasonable and efficient solution to a problem until you need to debug or extend it a few months later, and can't remember what it was supposed to be testing for.
- Libraries, especially open source ones, are great at isolating a small problem (like header parsing), anticipating edge cases, and testing them. Open source adds the benefit of community; other developers who depend on the library will discover bugs and contribute fixes, and we all reap the benefits of that effort without having to do any of that discovery, testing, or fixing ourselves.

I've added some thoughts on the current implementation, but I like where it's going.

**Comment by rrao24** [comment]
@a-type Thank you very much for your insight -- your comments make a lot of sense to me. I've updated the header parsing library so that it no longer mutates the parsed header object. Instead, a copy of the object is created, which contains all fields of the object except for `rel` and `url`. I also increased the maxdepth limit in .jshintrc to 4 as it was giving me linting errors for my updated code - I don't think the new code is poorly formatted, but please let me know if you feel differently. 

Could you all take another look at this PR whenever you get the chance? Thanks.

**Comment by dtolb** [comment]
@rrao24 please go ahead and merge [`v2.0-preview`](https://github.com/bandwidthcom/node-bandwidth/tree/2.0-preview) back into this branch.

**Comment by rrao24** [comment]
@dtolb Done.

**Comment by a-type** [comment]
@rrao24 to further clean up the code and reduce nesting depth, you probably want to use lodash's [omit](https://lodash.com/docs#omit) function. Lodash is a great tool with a lot of utility functions.

**Comment by rrao24** [comment]
@a-type Thank you for your feedback. I've rewritten the function to use lodash's omit method and have restored the .jshintrc file back to what it used to be. Lodash is now a dependency in package.json.

**Comment by a-type** [comment]
Great, thanks for responding to all the feedback! LGTM
